PurchasesCoreSwift/Networking/ETagManager.swift

```diff
@@ -0,0 +1,124 @@
+//
+// Created by César de la Vega on 4/16/21.
+//
+
+import Foundation
+
+@objc(RCETagManager) public class ETagManager: NSObject {
+    internal static let eTagHeaderName = "X-RevenueCat-ETag"
+
+    private let queue = DispatchQueue(label: "ETagManager")
+
+    private let userDefaults: UserDefaults
+
+    @objc public override init() {
+        self.userDefaults = UserDefaults(suiteName: ETagManager.suiteName) ?? UserDefaults.standard
+    }
+
+    @objc public init(userDefaults: UserDefaults) {
+        self.userDefaults = userDefaults
+    }
+
+    @objc public func eTagHeader(for urlRequest: URLRequest, refreshETag: Bool = false) -> [String: String] {
+        var storedETag = ""
+        if !refreshETag, let storedETagAndResponse = storedETagAndResponse(for: urlRequest) {
+            storedETag = storedETagAndResponse.eTag
+        }
+        return [ETagManager.eTagHeaderName: storedETag]
+    }
+
+    @objc public func getHTTPResultFromCacheOrBackend(with response: HTTPURLResponse,
+                                                      responseObject: [String: Any]?,
+                                                      error: Error?,
+                                                      request: URLRequest,
+                                                      retried: Bool) -> HTTPResponse? {
+        let statusCode = response.statusCode
+        let headersInResponse = response.allHeaderFields
+        let resultFromBackend = HTTPResponse(statusCode: statusCode, responseObject: responseObject)
+        guard error == nil else { return resultFromBackend }
+
+        let eTagInResponse: String? = headersInResponse[ETagManager.eTagHeaderName] as? String ??
+                headersInResponse[ETagManager.eTagHeaderName.lowercased()] as? String
+
+        if eTagInResponse != nil {
+            if shouldUseCachedVersion(responseCode: statusCode) {
+                guard let storedResponse = getStoredHTTPResponse(for: request) else {
```

Comment:
Yes, it's nil so that RCHTTPClient retries without etag because the cached version is corrupted. The backend result would have a 304 and be empty, so it wouldn't make sense to return that one.